File: golang/q03/part2/part2.go
package part2

import (
	c "aoc2023/q03/common"
	m "aoc2023/utils/math"
	"fmt"
)

func isNeighbour(part c.PartNumber, pos m.Vec2) bool {
	for y := pos.Y - 1; y <= pos.Y + 1; y++ {
		if y != part.NumberStart.Y {
			continue
		}

		for x := pos.X-1; x <= pos.X+1; x++ {
			if m.IsInBound(x, part.NumberStart.X, part.NumberEnd.X) {
				return true
			}
		}
	}

	return false
}

func Part2(schematic c.Schematic) {
	sum := 0

	for pos, symbol := range schematic.Symbols {
		if symbol != "*" {
			continue
		}

		neighbours := []c.PartNumber{}

		for _, part := range schematic.Numbers {
			if isNeighbour(part, pos) {
				neighbours = append(neighbours, part)
			}
		}

		if len(neighbours) != 2 {
			continue
		}

		power := neighbours[0].Value * neighbours[1].Value
		sum += power
	}

	fmt.Println("Part 2:", sum)
}

File: js/q03/part2/part2.js
import { parseFile } from "../common/common.js";

function isNeighbour(partNumber, pos) {
  const posParts = pos.split('-');
  const posX = parseInt(posParts[0]);
  const posY = parseInt(posParts[1]);

  for (let y = posY - 1; y <= posY + 1; y++) {
    if (y !== partNumber.start.y) {
      continue;
    }

    for (let x = posX - 1; x <= posX + 1; x++) {
      if (x >= partNumber.start.x && x <= partNumber.end.x) {
        return true;
      }
    }
  }

  return false;
}

function part2(path) {
  const parseResult = parseFile(path);
  if (parseResult.isErr()) {
    console.error(parseResult.unwrap());
    return;
  }

  let sum = 0;

  const schema = parseResult.unwrap();
  for (const entry of Object.entries(schema.symbols)) {
    const pos = entry[0];
    const symbol = entry[1];

    if (symbol !== '*') {
      continue;
    }

    const neighbours = [];

    for (const partNumber of schema.numbers) {
      if (isNeighbour(partNumber, pos)) {
        neighbours.push(partNumber);
      }
    }

    if (neighbours.length !== 2) {
      continue;
    }

    const power = parseInt(neighbours[0].number) * parseInt(neighbours[1].number);
    sum += power;
  }

  console.log(`Part 2: ${sum}`);
}

part2('./data/input.txt');
